exporter/otelarrowexporter/metadata.go

```diff
@@ -0,0 +1,201 @@
+	scf      streamClientFactory
+	host     component.Host
+
+	metadataKeys []string
+	exporters    sync.Map
+
+	// Guards the size and the storing logic to ensure no more than limit items are stored.
+	// If we are willing to allow "some" extra items than the limit this can be removed and size can be made atomic.
+	lock sync.Mutex
+	size int
+}
+
+var _ exp = (*metadataExporter)(nil)
+
+func newMetadataExporter(cfg component.Config, set exporter.Settings, streamClientFactory streamClientFactory) (exp, error) {
+	oCfg := cfg.(*Config)
+	// use lower-case, to be consistent with http/2 headers.
+	mks := make([]string, len(oCfg.MetadataKeys))
+	for i, k := range oCfg.MetadataKeys {
+		mks[i] = strings.ToLower(k)
+	}
+	sort.Strings(mks)
+	if len(mks) == 0 {
+		return newExporter(cfg, set, streamClientFactory)
+	}
+	return &metadataExporter{
+		config:       oCfg,
+		settings:     set,
+		scf:          streamClientFactory,
+		metadataKeys: mks,
+	}, nil
+}
+
+func (e *metadataExporter) getSettings() exporter.Settings {
+	return e.settings
+}
+
+func (e *metadataExporter) getConfig() component.Config {
+	return e.config
+}
+
+func (e *metadataExporter) helperOptions() []exporterhelper.Option {
+	return []exporterhelper.Option{
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		exporterhelper.WithTimeout(e.config.TimeoutSettings),
+		exporterhelper.WithRetry(e.config.RetryConfig),
+		exporterhelper.WithQueue(e.config.QueueSettings),
+		exporterhelper.WithStart(e.start),
+		exporterhelper.WithShutdown(e.shutdown),
+	}
+}
+
+func (e *metadataExporter) start(_ context.Context, host component.Host) (err error) {
+	e.host = host
+	return nil
+}
+
+func (e *metadataExporter) shutdown(ctx context.Context) error {
+	var err error
+	e.exporters.Range(func(_ any, value any) bool {
+		be, ok := value.(exp)
+		if !ok {
+			err = multierr.Append(err, fmt.Errorf("%w: %T", errUnexpectedType, value))
+			return true
+		}
+		err = multierr.Append(err, be.shutdown(ctx))
+		return true
+	})
+	return err
+}
+
+func (e *metadataExporter) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+	return be.pushTraces(ctx, td)
+}
+
+func (e *metadataExporter) pushMetrics(ctx context.Context, md pmetric.Metrics) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushMetrics(ctx, md)
+}
+
+func (e *metadataExporter) pushLogs(ctx context.Context, ld plog.Logs) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushLogs(ctx, ld)
+}
+
+func (e *metadataExporter) getOrCreateExporter(ctx context.Context, s attribute.Set, md metadata.MD) (exp, error) {
+	v, ok := e.exporters.Load(s)
+	if !ok {
+		e.lock.Lock()
```

Comment:
I would like to rewrite this function to de-indent the `if !ok` branch and use a defer function for the lock, so:

```
if ok {
  return v.(exp), nil
}
e.lock.Lock()
defer e.lock.Unlock()
// existing code w/o Unlock calls
return v.(exp), nil
```

Reply:
Fixed! Thanks for catching the failure to unlock in some of the return branches.